staking/contracts/Staking.sol

```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
 pragma abicoder v2;
 
 import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
-import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
 import "@rarible/lib-broken-line/contracts/LibBrokenLine.sol";
+import "@rarible/lib-broken-line/contracts/LibIntMapping.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "./INextVersionStake.sol";
 
-contract Staking {
+contract Staking is OwnableUpgradeable {
     using SafeMathUpgradeable for uint;
-    using LibBrokenLine for BrokenLineDomain.BrokenLine;
-
-    uint256 constant WEEK = 604800;                 //seconds one week
-    uint256 constant STARTING_POINT_WEEK = 2676;    //starting point week (Staking Epoch begining)
-    ERC20Upgradeable public token;
+    using LibBrokenLine for LibBrokenLine.BrokenLine;
+
+    uint256 constant WEEK = 604800;                         //seconds one week
+    uint256 constant STARTING_POINT_WEEK = 2676;            //starting point week (Staking Epoch begining)
+    uint256 constant TWO_YEAR_WEEKS = 104;                  //two year weeks
+    uint256 constant ST_FORMULA_MULTIPLIER = 1081000;       //stFormula multiplier = TWO_YEAR_WEEKS^2 * 100
+    uint256 constant ST_FORMULA_COMPENSATE = 1135050;       //stFormula compensate = (0.7+0.35) * ST_FORMULA_MULTIPLIER
+    uint256 constant ST_FORMULA_SLOPE_MULTIPLIER = 465;     //stFormula slope multiplier = 0.93 * 0.5 * 100
+    uint256 constant ST_FORMULA_CLIFF_MULTIPLIER = 930;     //stFormula cliff multiplier = 0.93 * 100
+
+    /**
+     * @dev ERC20 token to lock
+     */
+    IERC20Upgradeable public token;
+    /**
+     * @dev counter for locks identifiers
+     */
+    uint public counter;
+
+    /**
+     * @dev true if contract entered not working state
+     */
+    bool private stopped;
+    address public migrateTo;               //address migrate to
```

Comment:
I think, we need to disable stake and other functions when migration started

Reply:
Lets think:
migration  - long process and imagine: migration start, and for example ordinary life go on one or two weeks. 
If everything is all right with new contract(migrate to), and over 50% users already migrate so only after that set status stopped. Anyway we(contract owner) can call stop() before startMigration() we have this mechanism now, but
I think, we should not collect migration and stop in hard code.
My opinion, If @evgenynacu need, I do